src/H5TSpool.c

```diff
@@ -0,0 +1,386 @@
+
+    bool           shutdown;    /* Pool is shutting down */
+    unsigned       num_threads; /* # of threads in pool */
+    H5TS_thread_t *threads;     /* Array of worker threads in pool */
+
+    H5TS_pool_task_t *head, *tail; /* Task queue */
+};
+
+/********************/
+/* Local Prototypes */
+/********************/
+static herr_t                  H5TS__pool_free(H5TS_pool_t *pool);
+static H5TS_THREAD_RETURN_TYPE H5TS__pool_do(void *_pool);
+
+/*********************/
+/* Package Variables */
+/*********************/
+
+/*****************************/
+/* Library Private Variables */
+/*****************************/
+
+/*******************/
+/* Local Variables */
+/*******************/
+
+/* Declare a free list to manage H5TS_pool_t structs */
+H5FL_DEFINE_STATIC(H5TS_pool_t);
+
+/* Declare a free list to manage sequences of H5TS_thread_t structs */
+H5FL_SEQ_DEFINE_STATIC(H5TS_thread_t);
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS_pool_free
+ *
+ * Purpose:     Free resources for a thread pool, waiting for tasks to be invoked
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+static herr_t
+H5TS__pool_free(H5TS_pool_t *pool)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    /* Sanity check */
+    assert(pool);
+
+    /* Join all threads */
+    for (unsigned u = 0; u < pool->num_threads; u++)
+        if (H5_UNLIKELY(H5TS_thread_join(pool->threads[u], NULL) < 0))
+            HGOTO_DONE(FAIL);
+
+    /* Destroy the pool's mutex and condition variable */
+    if (H5_UNLIKELY(H5TS_mutex_destroy(&pool->mutex) < 0))
+        HGOTO_DONE(FAIL);
+    if (H5_UNLIKELY(H5TS_cond_destroy(&pool->cond) < 0))
+        HGOTO_DONE(FAIL);
+
+    /* Release memory */
+    if (pool->threads)
+        H5FL_SEQ_FREE(H5TS_thread_t, pool->threads);
+    H5FL_FREE(H5TS_pool_t, pool);
+
+done:
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__pool_free() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__pool_do
+ *
+ * Purpose:     Routine for worker threads to service tasks
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+static H5TS_THREAD_RETURN_TYPE
+H5TS__pool_do(void *_pool)
+{
+    H5TS_pool_t      *pool       = (H5TS_pool_t *)_pool; /* Pool for threads */
+    bool              have_mutex = false;                /* Whether we're holding the mutex */
+    H5TS_thread_ret_t ret_value  = (H5TS_thread_ret_t)0;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    /* Acquire tasks and invoke them, until pool is shut down */
+    while (1) {
+        /* Acquire the mutex for the pool */
+        if (H5_UNLIKELY(H5TS_mutex_lock(&pool->mutex) < 0))
+            HGOTO_DONE((H5TS_thread_ret_t)1);
```

Comment:
Should the return values in this function be -1 for failure instead of 1? Just going based off the values mentioned in the function header comment.

Reply:
It doesn't really matter, since the 'join' operation is only called when the pool is being freed and there's really nothing useful it can do if the thread failed.  This might be a situation where I would consider "poisoning" the library...

I'll switch it to '-1' though, for general consistency..